File: warranty-service/src/domain/CoverageScope.ts
import type { DefectCategory } from './DefectCategory.js';

/**
 * Value Object: CoverageScope
 * รายการ DefectCategory ที่ warranty ครอบคลุม
 */
export class CoverageScope {
  readonly categories: ReadonlySet<DefectCategory>;

  constructor(categories: Iterable<DefectCategory>) {
    const set = new Set(categories);
    if (set.size === 0) {
      throw new Error('CoverageScope requires at least one category');
    }
    this.categories = set;
  }

  includes(category: DefectCategory): boolean {
    return this.categories.has(category);
  }

  toArray(): DefectCategory[] {
    return [...this.categories];
  }
}

File: _java-archive/warranty-service/src/main/java/dev/kritchalach/warranty/domain/CoverageScope.java
package dev.kritchalach.warranty.domain;

import jakarta.persistence.Embeddable;
import jakarta.persistence.Enumerated;
import jakarta.persistence.EnumType;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.JoinColumn;

import java.util.HashSet;
import java.util.Set;

/**
 * Value Object: CoverageScope
 * รายการ DefectCategory ที่ warranty ครอบคลุม
 */
@Embeddable
public class CoverageScope {

    @ElementCollection(fetch = jakarta.persistence.FetchType.EAGER)
    @CollectionTable(name = "warranty_coverage_scope",
            joinColumns = @JoinColumn(name = "warranty_id"))
    @Enumerated(EnumType.STRING)
    @Column(name = "category")
    private Set<DefectCategory> categories = new HashSet<>();

    protected CoverageScope() {}

    public CoverageScope(Set<DefectCategory> categories) {
        if (categories == null || categories.isEmpty()) {
            throw new IllegalArgumentException("CoverageScope requires at least one category");
        }
        this.categories = new HashSet<>(categories);
    }

    public boolean includes(DefectCategory category) {
        return categories.contains(category);
    }

    public Set<DefectCategory> getCategories() {
        return Set.copyOf(categories);
    }
}
